meilisearch-types/src/keys.rs

```diff
@@ -328,44 +316,138 @@ impl Action {
     }
 
     pub const fn repr(&self) -> u8 {
-        *self as u8
+        self.bits()
+    }
+}
+
+impl<E: DeserializeError> Deserr<E> for Action {
+    fn deserialize_from_value<V: deserr::IntoValue>(
+        value: deserr::Value<V>,
+        location: deserr::ValuePointerRef<'_>,
+    ) -> Result<Self, E> {
+        match value {
+            deserr::Value::String(s) => match Self::ser_action_to_action(&s) {
+                Some(action) => Ok(action),
+                None => Err(deserr::take_cf_content(E::error::<std::convert::Infallible>(
+                    None,
+                    deserr::ErrorKind::Unexpected {
+                        msg: format!("string must be a valid action, got {}", s),
+                    },
+                    location,
+                ))),
+            },
+            _ => Err(take_cf_content(E::error(
+                None,
+                deserr::ErrorKind::IncorrectValueKind {
+                    actual: value,
+                    accepted: &[deserr::ValueKind::String],
+                },
+                location,
+            ))),
+        }
+    }
+}
+
+impl Serialize for Action {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(Self::action_to_ser_action(self))
+    }
+}
+
+impl<'de> Deserialize<'de> for Action {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct Visitor;
+        impl<'de> serde::de::Visitor<'de> for Visitor {
+            type Value = Action;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(formatter, "the name of a valid action (string)")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                match Self::Value::ser_action_to_action(s) {
+                    Some(action) => Ok(action),
+                    None => Err(E::invalid_value(serde::de::Unexpected::Str(s), &"a valid action")),
+                }
+            }
+        }
+
+        deserializer.deserialize_str(Visitor)
+    }
+}
+
+impl Sequence for Action {
+    const CARDINALITY: usize = Self::SERDE_MAP_ARR.len();
+
+    fn next(&self) -> Option<Self> {
+        let next_index = self.bits() as usize + 1;
+        if next_index == Self::CARDINALITY {
+            None
+        } else {
+            Some(Self::SERDE_MAP_ARR[next_index].1)
+        }
+    }
+
+    fn previous(&self) -> Option<Self> {
+        if self.bits() == 0 {
+            None
+        } else {
+            Some(Self::SERDE_MAP_ARR[self.bits() as usize - 1].1)
+        }
+    }
+
+    fn first() -> Option<Self> {
+        Some(Self::SERDE_MAP_ARR[0].1)
+    }
+
+    fn last() -> Option<Self> {
+        Some(Self::SERDE_MAP_ARR[Self::CARDINALITY - 1].1)
     }
 }
```

Comment:
I'm also not liking this very much, how could I improve it?

Reply:
What are you talking about? This comment is floating around for me 😂

Reply:
Well, I'm just not entirely sure if this is the best way to implement `Sequence` for this. Thought I'd get some input.